rfcs/experimental/flow_graph_wait_for_one/README.md

```diff
@@ -0,0 +1,481 @@
+Metainformation object copy is submitted together with each element into each output port.
+
+Metainfo should not be split since is is unclear what is the relation between elements in the tuple and the metainformation objects stored in the internal list so all of them should
+be provided to the successors.
+
+### ``indexer_node``
+
+``indexer_node`` only tags the input value and broadcasts it with the original metainfo to the successors.
+
+### ``composite_node``
+
+``composite_node`` does not require any additional changes. `try_put_and_wait` feature and the explicit support for metainformation should be done by the nodes inside of the composite.
+
+## Additional implementation details
+
+### Details about metainformation class
+
+``message_metainfo`` class synopsis:
+
+```cpp
+class message_metainfo
+{
+public:
+    using waiters_type = std::forward_list<d1::wait_context_vertex*>;
+
+    message_metainfo();
+
+    message_metainfo(const message_metainfo&);
+    message_metainfo(message_metainfo&&);
+
+    message_metainfo& operator=(const message_metainfo&);
+    message_metainfo& operator=(message_metainfo&&);
+
+    const waiters_type& waiters() const &;
+    waiters_type&& waiters() &&;
+
+    bool empty() const;
+
+    void merge(const message_metainfo&);
+};
+```
+
+The current implementation of ``message_metainfo`` class wraps only the list of single message waiters. The class may be extended if necessary to cover additional use-cases.
+
+The metainfo is required to hold a list of message waiters instead of single waiter to cover the ``continue_node`` and ``join_node`` joining use-cases. Consider the example:
+
+```cpp
+using namespace oneapi::tbb;
+
+flow::function_node<int, int> start1(g, ...);
+flow::function_node<int, int> start2(g, ...);
+
+flow::join_node<std::tuple<int, int>> join(g);
+
+flow::function_node<std::tuple<int, int>, int> post_process(g, ...);
+
+flow::make_edge(start, flow::input_port<0>(join));
+flow::make_edge(start, flow::input_port<1>(join));
+flow::make_edge(join, post_process);
+
+std::thread t1([&]() {
+    start1.try_put_and_wait(1);
+});
+
+std::thread t2([&]() {
+    start2.try_put_and_wait(2);
+})
+```
+
+### Combined or separated wait
+
+Current proposal describes only the case where submitting the work into the flow graph and waiting for it are combined in a single use API `node.try_put_and_wait`.
+In theory, it can be useful to also have the ability to split these phases:
+
+```cpp
+
+oneapi::tbb::flow::graph g;
+oneapi::tbb::flow::function_node<int, int> start_node(g, ...);
+
+auto desc = node.try_put(work); // other API name can be selected
+
+// Other work
+// May submit more work into the Flow Graph
+// May create other descriptors that would be waited later
+
+g.wait(desc);
+
+```
+
+In that case it would be needed to extend the node with the new API returning some descriptor that can be used as a hit work waiting function `g.wait` that also should be added.
```

Comment:
```suggestion
In that case it would be needed to extend the node with the new API returning some descriptor that can be used as the argument to the work waiting function `g.wait` that also should be added.
```